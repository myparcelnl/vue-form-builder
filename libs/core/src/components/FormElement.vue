<template>
  <component
    :is="element.component"
    v-model="model"
    v-bind="attributes"
    v-on="hooks">
    <slot />
  </component>
</template>

<script lang="ts">
import {PropType, computed, defineComponent, provide} from 'vue';
import {AnyElementInstance} from '../types';
import {INJECT_ELEMENT} from '../services';
import {useElementHooks} from '../composables';

export default defineComponent({
  name: 'FormElement',
  props: {
    element: {
      type: Object as PropType<AnyElementInstance>,
      required: true,
    },
  },

  setup: (props) => {
    provide(INJECT_ELEMENT, props.element);

    return {
      hooks: useElementHooks(props.element, {
        blur: props.element.blur,
        focus: props.element.focus,
        click: props.element.click,
      }),

      model: computed({
        get() {
          return props.element.ref;
        },
        set(value) {
          // eslint-disable-next-line vue/no-mutating-props
          props.element.ref = value;
        },
      }),

      /**
       * Collect attributes. Always adds `element.attributes`, but only adds `element` and `element.props` if element is a Vue component.
       */
      attributes: computed(() => {
        return {
          ...props.element.attributes,
          ...(typeof props.element.component === 'string'
            ? {}
            : {
                ...props.element.props,
                element: props.element,
              }),
        };
      }),
    };
  },
});
</script>
