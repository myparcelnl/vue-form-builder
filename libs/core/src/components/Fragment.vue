<template>
  <template v-if="component">
    <component :is="component">
      <slot />
    </component>
  </template>
  <slot v-else />
</template>

<script lang="ts">
import {type PropType, defineComponent} from 'vue';
import {type ComponentOrHtmlElement} from '../types';

export default defineComponent({
  name: 'Fragment',
  props: {
    component: {
      type: [String, Object] as PropType<ComponentOrHtmlElement>,
      default: null,
    },
  },
});
</script>
