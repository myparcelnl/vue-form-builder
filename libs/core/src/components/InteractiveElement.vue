<template>
  <div
    v-show="element.isVisible"
    :id="elementId"
    :class="element.isVisible ? element.form.config.fieldClass : null">
    <component
      :is="element.component"
      :id="element.name ?? element.name"
      v-model="model"
      :label="element.label"
      :name="element.name"
      :errors="element.formattedErrors"
      :disabled="element.isDisabled"
      :valid="element.isValid"
      :suspended="element.isSuspended"
      v-bind="{...$attrs, ...element.props}"
      v-on="hooks" />
  </div>
</template>

<script lang="ts">
import {PropType, computed, defineComponent, provide} from 'vue';
import {INJECT_ELEMENT} from '../services';
import {InteractiveElementInstance} from '../form';
import {useLifeCycleHooks} from '../composables';

export default defineComponent({
  name: 'InteractiveElement',
  inheritAttrs: false,
  props: {
    element: {
      type: Object as PropType<InteractiveElementInstance>,
      required: true,
    },
  },

  setup: (props) => {
    const lifecycleHooks = useLifeCycleHooks();
    provide(INJECT_ELEMENT, props.element);

    lifecycleHooks.register(props.element.hooks, props.element);

    const hooks = computed(() => {
      const registeredHooks = props.element.hooks.getRegisteredHooks();

      return registeredHooks.reduce(
        (acc, hook) => {
          if (!hook.name.startsWith('on')) {
            return acc;
          }

          return {
            ...acc,
            // todo: fix types
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            [hook.name.replace(/^on/, '')]: (value: any) => hook.callback(props.element, value),
          };
        },
        {
          blur: props.element.blur,
          focus: props.element.focus,
          click: props.element.click,
        },
      );
    });

    return {
      hooks,
      model: computed({
        get() {
          return props.element.ref;
        },
        set(value) {
          // eslint-disable-next-line vue/no-mutating-props
          props.element.ref = value;
        },
      }),

      elementId: computed(() => (props.element.name ? `${props.element.name}__wrapper` : '')),
    };
  },
});
</script>
