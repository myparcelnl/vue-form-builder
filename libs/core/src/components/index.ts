export {default as MagicForm} from './MagicForm.vue';
