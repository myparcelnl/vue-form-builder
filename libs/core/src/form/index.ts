export * from './Form';
export * from './Form.types';
export * from './defaultFormConfiguration';
export * from './defineField';
export * from './defineForm';
export * from './interactive-element';
export * from './plain-element';
export * from './validator';
