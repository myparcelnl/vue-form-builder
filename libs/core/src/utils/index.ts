export * from './getValue';
export * from './markComponentAsRaw';
export * from './useDynamicWatcher';
