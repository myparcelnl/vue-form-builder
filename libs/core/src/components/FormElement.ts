import {Component, PropType, Teleport, defineComponent, h} from 'vue';
import {FormInstance, InteractiveElementInstance} from '../form';
import {AnyElementInstance} from '../types';
import InteractiveElement from './InteractiveElement.vue';
import PlainElement from './PlainElement.vue';
import {isOfType} from '@myparcel/ts-utils';
import {useTestAttributes} from '../composables/useTestAttributes';

// noinspection JSUnusedGlobalSymbols
export default defineComponent({
  name: 'FormElement',
  props: {
    form: {
      type: Object as PropType<FormInstance>,
      required: true,
    },
    element: {
      type: Object as PropType<AnyElementInstance>,
      required: true,
    },
  },

  render() {
    let component: Component = h(
      isOfType<InteractiveElementInstance>(this.element, 'ref') ? InteractiveElement : PlainElement,
      {
        ...useTestAttributes(this.element),
        element: this.element,
      },
    );

    const hasOwnWrapper = typeof this.element.wrapper !== 'boolean';
    const inheritsWrapper = this.element.wrapper === true && this.form.config.field?.wrapper;

    if (inheritsWrapper || hasOwnWrapper) {
      const childComponent = component;

      component = h(
        (hasOwnWrapper ? this.element.wrapper : this.form.config.field?.wrapper) as Component,
        {
          ...this.$attrs,
          ...this.form.config.fieldDefaults.attributes,
          element: this.element,
        },
        () => [childComponent],
      );
    }

    if (this.element.teleportSelector) {
      // @ts-expect-error Teleport has an extra property which doesn't fit in the Component type
      return h(Teleport, {to: this.element.teleportSelector}, [component]);
    }

    return component;
  },
});
