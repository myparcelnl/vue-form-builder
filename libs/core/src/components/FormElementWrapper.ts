import {Component, PropType, Teleport, defineComponent, h} from 'vue';
import {AnyElementInstance} from '../types';
import FormElement from './FormElement.vue';
import {FormInstance} from '../form';
import {useTestAttributes} from '../composables';

// noinspection JSUnusedGlobalSymbols
export default defineComponent({
  name: 'FormElementWrapper',
  props: {
    form: {
      type: Object as PropType<FormInstance>,
      required: true,
    },
    element: {
      type: Object as PropType<AnyElementInstance>,
      required: true,
    },
  },

  render() {
    let component: Component = h(FormElement, {
      ...useTestAttributes(this.element),
      element: this.element,
    });

    const hasOwnWrapper = typeof this.element.wrapper !== 'boolean';
    const inheritsWrapper = this.element.wrapper === true && this.form.config.field?.wrapper;

    if (inheritsWrapper || hasOwnWrapper) {
      const childComponent = component;

      component = h(
        (hasOwnWrapper ? this.element.wrapper : this.form.config.field?.wrapper) as Component,
        {
          ...this.$attrs,
          ...this.form.config.fieldDefaults.attributes,
          element: this.element,
        },
        () => [childComponent],
      );
    }

    if (this.element.teleportSelector) {
      // @ts-expect-error Teleport has an extra property which doesn't fit in the Component type
      return h(Teleport, {to: this.element.teleportSelector}, [component]);
    }

    return component;
  },
});
