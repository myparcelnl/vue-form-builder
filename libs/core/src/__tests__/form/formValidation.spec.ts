import {
  canNotContainLetterValidator,
  firstNameNotJohnValidator,
  nameNotDonald,
  nameNotDonaldMack,
} from './interactive-element/validationData';
import {canNotContainX, firstNameNotDuane} from '../examples/validators';
import {defineField, defineForm} from '../../form';
import {describe, expect, it} from 'vitest';
import {flushPromises, mount} from '@vue/test-utils';
import {formIsInvalid, formIsValid, generateForm} from '../utils';
import {MagicForm} from '../../components';
import SubmitButton from '../elements/SubmitButton.vue';
import TextInput from '../elements/TextInput.vue';
import {ref} from 'vue';

describe('Form and field validation', () => {
  it('can determine if a text input is valid based on single predicate', async () => {
    expect.assertions(11);
    const firstName = ref('');
    const lastName = ref('');
    const validationForm = defineForm('validationForm', {
      fields: [
        defineField({
          name: 'firstName',
          component: TextInput,
          ref: firstName,
          validate: (_, value) => String(value).startsWith('J'),
          errorMessage: 'Field must start with "J"',
        }),
        defineField({
          name: 'lastName',
          component: TextInput,
          ref: lastName,
        }),
        defineField({
          component: SubmitButton,
        }),
      ],
    });

    const wrapper = mount(MagicForm, {props: {form: validationForm}});
    await flushPromises();

    // expect default state to be valid regardless of input
    formIsValid(validationForm);

    await wrapper.find('input[name="firstName"]').setValue('Karen');
    expect(firstName.value).toBe('Karen');

    await wrapper.find('input[name="lastName"]').setValue('Doe');
    expect(lastName.value).toBe('Doe');

    await validationForm.submit();

    const firstNameWarning = wrapper.findByTest({id: 'firstName', type: 'errors'});
    expect(firstNameWarning.exists()).toBe(true);
    expect(firstNameWarning.text()).toBe('Field must start with "J"');

    const lastNameWarning = wrapper.findByTest({id: 'firstName', type: 'errors'});
    expect(lastNameWarning.exists()).toBe(false);

    formIsInvalid(validationForm);
  });

  it('can determine if a text input is valid based on previous inputs and predicates', async () => {
    const firstName = ref('');
    const lastName = ref('');
    const validationForm = defineForm('validationForm', {
      fields: [
        defineField({
          name: 'firstName',
          component: TextInput,
          ref: firstName,
          validate: (_, value) => String(value).startsWith('J'),
          errorMessage: 'Field must start with "J"',
        }),
        defineField({
          name: 'lastName',
          component: TextInput,
          ref: lastName,
          validate: (field, value) => !(field.form.model.firstName.ref.value === 'Jack' && String(value) === 'McGill'),
          errorMessage: 'Last name cannot be "McGill" if first name is "Jack"',
        }),
        defineField({
          component: SubmitButton,
        }),
      ],
    });
    const wrapper = mount(MagicForm, {props: {form: validationForm}});
    await flushPromises();
    formIsValid(validationForm);

    await wrapper.findByTest({id: 'firstName', type: 'interactive'}).setValue('Jack');
    await wrapper.findByTest({id: 'lastName', type: 'interactive'}).setValue('McGill');

    await validationForm.submit();
    expect(wrapper.html()).toMatchSnapshot();
    const lastNameWarning = wrapper.findByTest({id: 'lastName', type: 'errors'});
    expect(lastNameWarning.text()).toBe('Last name cannot be "McGill" if first name is "Jack"');

    formIsInvalid(validationForm);

    await wrapper.find('input[name="lastName"]').setValue('Taves');

    await validationForm.submit();
    formIsValid(validationForm);
  });

  it('validates using a single function', async () => {
    expect.assertions(4);

    const form = generateForm([
      defineField({
        component: 'input',
        name: 'element',
        ref: ref(''),
        validate: (_, value) => String(value).startsWith('J'),
        errorMessage: 'Field must start with "J"',
      }),
    ]);

    form.model.element.ref.value = 'Peter';
    await form.submit();
    expect(form.isValid.value).toBe(false);
    expect(form.model.element.errors.value).toEqual(['Field must start with "J"']);

    form.model.element.ref.value = 'Joe Mater';
    await form.submit();
    expect(form.isValid.value).toBe(true);
    expect(form.model.element.errors.value).toEqual([]);
  });

  it.skip('moves the error message to another field', async () => {
    // TODO: fix this unit test to properly deal with form.model.target.errors with .value
    expect.assertions(6);

    const form = generateForm([
      defineField({
        component: 'div',
        name: 'target',
      }),
      defineField({
        component: 'input',
        name: 'element',
        ref: ref(''),
        validate: (_, value) => String(value).startsWith('J'),
        errorMessage: 'Field must start with "J"',
        errorsTarget: 'target',
      }),
    ]);

    form.model.element.ref.value = 'Peter';
    await form.submit();
    await flushPromises();
    expect(form.isValid.value).toBe(false);
    expect(form.model.target.errors.value).toEqual(['Field must start with "J"']);
    expect(form.model.element.errors.value).toEqual([]);

    form.model.element.ref.value = 'Joe Mater';
    await form.submit();
    await flushPromises();
    expect(form.isValid.value).toBe(true);
    expect(form.model.target.errors.value).toEqual([]);
    expect(form.model.element.errors.value).toEqual([]);
  });

  it('validates using an array of validators', async () => {
    expect.assertions(1);

    const form = generateForm({
      fields: [
        defineField({
          component: 'input',
          name: 'element',
          ref: ref('Jack'),
          validators: [firstNameNotJohnValidator(), canNotContainLetterValidator()],
        }),
      ],
    });

    await form.submit();
    expect(form.isValid.value).toBe(true);
  });

  it('validates using a computed validator', async () => {
    expect.assertions(1);

    const form = generateForm({
      fields: [
        defineField({
          component: 'input',
          name: 'element',
          ref: ref('Jack'),
          validators: [firstNameNotDuane(), canNotContainX()],
        }),
      ],
    });

    await form.submit();
    expect(form.isValid.value).toBe(true);
  });

  it('validates using an array of validators, with precedence', async () => {
    expect.assertions(3);

    const form = generateForm({
      fields: [
        defineField({
          component: 'input',
          name: 'element',
          ref: ref('Jack'),
          validators: [
            // validators without precedence:
            firstNameNotJohnValidator(),
            canNotContainLetterValidator(),
            // validators with precedence:
            nameNotDonaldMack(1),
            nameNotDonald(2),
          ],
        }),
      ],
    });

    await form.submit();
    expect(form.isValid.value).toBe(true);
    form.model.element.ref.value = 'Donald Mack';
    await form.submit();
    expect(form.model.element.errors.value).toEqual(['Donald Mack, we specifically do not send to you.']);
    form.model.element.ref.value = 'Donald';
    await form.submit();
    expect(form.model.element.errors.value).toEqual(['Donald, we do not send to you.']);
  });
});
