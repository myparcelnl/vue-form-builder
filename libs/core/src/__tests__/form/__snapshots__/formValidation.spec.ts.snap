// Vitest Snapshot v1

exports[`Form and field validation > can determine if a text input is valid based on previous inputs and predicates 1`] = `"<form id=\\"validationForm\\" class=\\"invalid\\"><input id=\\"firstName\\" type=\\"text\\" name=\\"firstName\\" class=\\"\\" data-test-id=\\"firstName\\" data-test-type=\\"interactive\\"><input id=\\"lastName\\" type=\\"text\\" name=\\"lastName\\" class=\\"border-red-500\\" data-test-id=\\"lastName\\" data-test-type=\\"interactive\\"><button type=\\"submit\\" data-test-type=\\"plain\\">Submit</button></form>"`;
